### DISCUSSION TITLE
Create a fallback for lazy routes

### DISCUSSION THREAD

**Discussion by ryan-saffer** [remix-run/react-router#10540] [Proposals]
### Background
Since the release of a `lazy` route in 6.9, there is no way to provide a fallback component to display while the route is being loaded.

It seems the only way to provide one would be to not use react-routers `lazy` property, but instead use `React.lazy` to wrap our `element`  in a `React.Suspense`. Without the ability to do this using react-routers `lazy` property, I don't really see a benefit to using it over `React.lazy`?

### Using react-router-dom
```typescript
// When doing this, there is no way to provide a fallback.
// When I click on a link that takes me to `my-route`, the app appears to freeze for a moment as it loads the component.
const router = createBrowserRouter([
    {
        path: 'my-route',
        lazy: () => import('./MyComponent'),
    },
])
```

### Using `React.Suspense`
```typescript
const MyComponent = React.lazy(() => import('./MyComponent'))

// but now I'm not using the 'lazy' property of the route..
const router = createBrowserRouter([
    {
        path: 'my-route',
        element:
            <React.Suspense fallback={<Loading />}>
                <MyComponent />
            <React.Suspense />
    }
])
```

Could anyone help me understand why I would use the first example instead of the second?

**Comment by brophdawg11** [comment]
It's a UX decision at the end of the day

* `React.lazy` is for render-then-fetch - absolutely valid to keep using that if that's your desired UX
* `route.lazy` is for fetch-then-render - which is not quite possible with `React.lazy` + Data Routers (or at least is [overly complicated](https://www.infoxicator.com/en/react-router-6-4-code-splitting))

I personally prefer the latter to continue showing the user content during loading instead of flipping them to skeletons

**Reply by paolostyle** [reply]
I'm not sure if you'll be able to advise here, but how do I ensure the users still get some feedback when navigating to a route loaded with `route.lazy`? You're saying you "prefer to continue showing the user content during loading", but my experience is that it feels like the app is unresponsive. Imagine a simple scenario, a route with a sidebar that contains NavLinks to subroutes ( and an `<Outlet />`. Each subroute is loaded with `route.lazy`. Some chunks might be quite large and it might take e.g. a second to load one. How do I give the user instant feedback when they click the link ("Hey, we're loading the page, don't worry")? Currently, it's either impossible or I don't know how to do it. 

If there isn't - out of curiosity, can you provide an example where such behavior is generally acceptable? I acknowledge that this might be highly subjective, even if I'm quite confident that some form of instant feedback is the generally desired behavior for any UI interaction (e.g. if a user clicks a link on a regular, non-SPA page, there is a spinner in the browser tab).

**Reply by brophdawg11** [reply]
Definitely not impossible 😉 

We do exactly this on our own [docs](https://remix.run/docs) where we slightly fade out the current content and show a status bar at the top while content loads (throttle to 3G to see it best).

The hook you're looking for is [`useNavigation`](https://reactrouter.com/en/main/hooks/use-navigation) which will be in a non-`idle` status during navigations, which includes the resolution of `route.lazy`.

**Reply by paolostyle** [reply]
Awesome, thanks so much. I will try it but looking at the docs this is exactly what I needed.

**Reply by einarq** [reply]
> Definitely not impossible 😉
> 
> We do exactly this on our own [docs](https://remix.run/docs) where we slightly fade out the current content and show a status bar at the top while content loads (throttle to 3G to see it best).
> 
> The hook you're looking for is [`useNavigation`](https://reactrouter.com/en/main/hooks/use-navigation) which will be in a non-`idle` status during navigations, which includes the resolution of `route.lazy`.

@brophdawg11 
Could you point me to an example of how to implement this type of loading state UI? Is there an example in your code somewhere?
I'm struggling a lot with loaders and lazy routes and error handling etc with React Router unfortunately. Not finding the api very easy to groc.

**Reply by brophdawg11** [reply]
Can you elaborate on what you're struggling with?  [`navigation.state`](https://reactrouter.com/en/main/hooks/use-navigation#navigationstate) tells you the navigation state so you can use that to show a spinner or a progress bar.

```jsx
// potentially in your app layout
function Layout() {
  let navigation = useNavigation();
  return (
    <div>
      {navigation.state !== "idle" ? <Spinner /> : null}
      <Header />
      <Outlet />
      <Footer />
    </div>
  );  
}
```

If you are having specific issues, it would be helpful if you could provide a slimmed down code sandbox example what demonstrates the problem.

**Reply by einarq** [reply]
That works on the top-level, but harder to achieve on a sub-route. I guess you then have to use some type of logic to check every to and from location, to check where to display the loading dots.

Disclaimer: I know that loading dots is not the best UX, but its what we have. I will probably try to convert to a "loading bar+gray out content" approach and see what people think.

**Reply by VillePakarinenPosti** [reply]
@brophdawg11 Could `HydrateFallback` prop on `route` be extended so that it would display on subsequent navigations if the matched path has unresolved `lazy` or the upcoming `patch-route` defined?

**Reply by Hypermona** [reply]
> Can you elaborate on what you're struggling with? [`navigation.state`](https://reactrouter.com/en/main/hooks/use-navigation#navigationstate) tells you the navigation state so you can use that to show a spinner or a progress bar.
> 
> ```js
> // potentially in your app layout
> function Layout() {
>   let navigation = useNavigation();
>   return (
>     <div>
>       {navigation.state !== "idle" ? <Spinner /> : null}
>       <Header />
>       <Outlet />
>       <Footer />
>     </div>
>   );  
> }
> ```
> 
> If you are having specific issues, it would be helpful if you could provide a slimmed down code sandbox example what demonstrates the problem.

Instead of this why can't we accept the fallback component and handle it internally

**Comment by HazyFish** [comment]
`useNavigation()` gives the navigation state of the entire app. If I have deeply nested routes and use `useNavigation()` on top level layout component, then when user navigates between sub-sub-routes of the same sub-route, we are showing the fallback at top level. It does not solve my problem either.

**Reply by einarq** [reply]
This is definitely an issue, agree:
https://github.com/remix-run/react-router/discussions/10540#discussioncomment-9730255

**Reply by einarq** [reply]
So I wouldn’t show a spinner or something that triggers unmount, but rather some type of transparent overlay
